### PR TITLE
support s3 pattern hourly, minitely, secondly partitioned dataset ingestion

### PR DESCRIPTION
**Pull request by chenqin**
used to have specific discussion

### CODE REVIEW

File: src/ingest/SpecRepo.cpp
Comment (shawncao) [inline src/ingest/SpecRepo.cpp:157]: do you need padding to ensure two digit literals like `hour=01`?

File: src/ingest/SpecRepo.cpp
Comment (shawncao) [inline src/ingest/SpecRepo.cpp:184]: I feel we may need to introduce a way to tell how many different macros are needed to gen file specs.

s3://zzz/%date%/ => [date]
s3://zzz/ds=%date%/hr=%hour%/ => [date, hour]

And we can maintain a list of macro definition internally, this needs to be refactored from single date value into a property bag or an object definition to support all macros:  https://github.com/varchar-io/nebula/blob/2303e4f11e40eca1566ed20c83fd5223a2efdc9d/src/ingest/IngestSpec.h#L68

File: src/ingest/SpecRepo.cpp
Comment (shawncao) [inline src/ingest/SpecRepo.cpp:153]: update the timeValue to reflect the hour value.

(Note, this is only useful when use macro value as time value, if time value is coming from a column, this value is useless)

File: src/ingest/SpecRepo.cpp
Comment (shawncao) [reply src/ingest/SpecRepo.cpp:184]: And based on different macro set - we may have different loops.

File: src/ingest/SpecRepo.cpp
Comment (shawncao) [inline src/ingest/SpecRepo.cpp:182]: sounds great - wondering what's the best way to plug it in. If you target this, could you please file an issue for it?

**Review by shawncao — COMMENTED** [review]
Amazingly you found the code spot purely by yourself!
I added some comments - code format, refactoring macro support is flexible which could be done separately if we want to keep improving this part.

File: src/ingest/SpecRepo.cpp
Comment (chenqin) [reply src/ingest/SpecRepo.cpp:179]: yup, ctrl+alt+L works great!

Comment (shawncao) [reply src/ingest/SpecRepo.cpp:179]: I'm guessing you're using Microsoft Ergo Keyboard? 😉

File: src/ingest/SpecRepo.cpp
Comment (chenqin) [reply src/ingest/SpecRepo.cpp]: genSpecs4HourlyRoll looks hacky, I updated pr with date/hour/minute/second granularity bucketing.
Ultimately, I am targeting to 
- scan and ingest smaller granularity bucket data into nebula (e.g minute level) periodically. 
- (pending discussion) offload one granularity higher parquet /csv/ orc file to upper granularity bucket (e.g hourly level) and put a _SUCCESS file when it complete.

Comment (shawncao) [reply src/ingest/SpecRepo.cpp]: I'm looking at the new changes now - I'm assuming you meant some external process to do the work for your point number 2 listed above? Or you meant some internal auto roll-up process to transform lower granularity data to higher granularity? 

(internal roll-up could be a very interesting feature to have in Nebula but it's a big project deserved a nice design)

File: src/ingest/IngestSpec.h
Comment (shawncao) [inline src/ingest/IngestSpec.h:182]: this is not used and not complied with member naming convention. 

BTW, is this PR ready for review? Would love to wait for it's ready, so we can skip these small things if it's still in draft as I noticed in the PR labels.

File: src/ingest/SpecRepo.cpp
Comment (shawncao) [inline src/ingest/SpecRepo.cpp:82]: 👍

File: src/meta/TableSpec.h
Comment (shawncao) [inline src/meta/TableSpec.h:220]: would rather to have a static hash map to store this mapping. 
nebula:common::unordered_map is super fast - defined by https://github.com/varchar-io/nebula/blob/master/src/common/Hash.h#L169

File: src/ingest/TimeRow.h
Comment (shawncao) [inline src/ingest/TimeRow.h:146]: nit: this could be renamed too

File: src/ingest/SpecRepo.cpp
Comment (shawncao) [inline src/ingest/SpecRepo.cpp:169]: I suggest we split each pattern type process into a subroutine instead of in a multi layers of `for blocks`.
Clearly to declare that we only supports 4 different patterns for now, even though we may evolve into more generic framework.

In addition to that, we may consider supporting categorical pattern as well for partitioning by enumerable values,  such as `s3://zzz/app=[a, b, c]/dt={date}/hour={hr}`, in this case, app will have 3 different values `a` or `b` or `c`

File: src/ingest/SpecRepo.cpp
Comment (chenqin) [reply src/ingest/SpecRepo.cpp]: linux

File: src/ingest/SpecRepo.cpp
Comment (chenqin) [reply src/ingest/SpecRepo.cpp]: actually, let's just keep using timeValue as timestamp for now. We might looks into support timestamp extraction from column if needed.

File: src/ingest/SpecRepo.cpp
Comment (chenqin) [reply src/ingest/SpecRepo.cpp]: We can start leverage external system and schedule corn jobs for now.
nebula customized roll-up indeed interesting from realtime serving->batch ingestion prospective. Most naive use case can be we have 5 mins partitioned data partitions ingested to nebula, nebula upload compacted parquet files to a hourly directory once observed timeValue greater than give hour.

File: src/ingest/IngestSpec.h
Comment (chenqin) [reply src/ingest/IngestSpec.h:182]: fixed, still need to add tests.

File: src/meta/TableSpec.h
Comment (chenqin) [reply src/meta/TableSpec.h:220]: fixed.

File: src/ingest/SpecRepo.cpp
Comment (chenqin) [reply src/ingest/SpecRepo.cpp]: Updated and remove for loops with iterative approach. Still haven't got chance working on app=[a, b, c] like macro due to the fact we need to support predefined collection macro pattern or on demand list based collection generation requires a bit more work

File: src/ingest/IngestSpec.h
Comment (shawncao) [inline src/ingest/IngestSpec.h:59]: are these string literals used? If yes, use `static constexpr` instead

File: src/ingest/TimeRow.h
Comment (shawncao) [reply src/ingest/TimeRow.h:146]: I meant mdate - could you rename it to watermark as well?

File: src/meta/TableSpec.h
Comment (shawncao) [inline src/meta/TableSpec.h:211]: to be consistently as `static const`

also could you please wrap each entry into one line? (we want to enforce max line width some day with build rules)

File: src/meta/TableSpec.h
Comment (shawncao) [inline src/meta/TableSpec.h:217]: const std::string&

File: src/ingest/SpecRepo.cpp
Comment (shawncao) [inline src/ingest/SpecRepo.cpp:149]: do not start with _ for local var?

File: src/meta/TableSpec.h
Comment (shawncao) [reply src/meta/TableSpec.h:211]: We should use UPPER_CASE for static const values, especially namespace level.

Comment (shawncao) [reply src/meta/TableSpec.h:211]: I think 'date' literal can't be macro, the real macro defined in Nebula already is `{date}`,  it's `%7Bdate%7D` after url encoding.

File: src/meta/TableSpec.h
Comment (shawncao) [inline src/meta/TableSpec.h:219]: this doesn't look right - we should check the result against `std::string::npos` which is effectively -1

File: src/meta/TableSpec.h
Comment (shawncao) [inline src/meta/TableSpec.h:220]: to prevent repeatedly `find` calls, can we introduce local variables for them? 
```
const auto tsMacroFound = pattern.find(MacroPattern.at(PatternMacro::TIMESTAMP)) != std::string::npos;
const auto dateMacroFound = pattern.find(MacroPattern.at(PatternMacro::DATE)) != std::string::npos;
const auto hourMacroFound = pattern.find(MacroPattern.at(PatternMacro::HOUR)) != std::string::npos;
const auto minuteMacroFound = pattern.find(MacroPattern.at(PatternMacro::MINUTE)) != std::string::npos;
...
if(xxFond){
 ..
}
else if(yyFound && ...){
  ...
}
```

This will make the code a bit more readable.

File: src/meta/TableSpec.h
Comment (shawncao) [reply src/meta/TableSpec.h:217]: I think we should add a unit test for this function.

File: src/meta/TableSpec.h
Comment (shawncao) [inline src/meta/TableSpec.h:19]: I may be wrong but I don't think this is used.

File: src/meta/TableSpec.h
Comment (shawncao) [inline src/meta/TableSpec.h:214]: I suggest using these names: `ChildPattern`, `UnitInSeconds` and `ChildSize`. (do not use `partition` to confuse other definitions of partition)

**Review by shawncao — CHANGES_REQUESTED** [review]
Thanks for the update, logically looks good to me. But I have some comments requiring updates before approval.

File: src/ingest/IngestSpec.h
Comment (chenqin) [reply src/ingest/IngestSpec.h:59]: removed

File: src/meta/TableSpec.h
Comment (chenqin) [reply src/meta/TableSpec.h:219]: removed

File: src/meta/TableSpec.h
Comment (chenqin) [reply src/meta/TableSpec.h:19]: removed

File: src/meta/TableSpec.h
Comment (chenqin) [reply src/meta/TableSpec.h:219]: removed

File: src/ingest/SpecRepo.cpp
Comment (chenqin) [reply src/ingest/SpecRepo.cpp:149]: fixed

File: src/ingest/IngestSpec.h
Comment (chenqin) [reply src/ingest/IngestSpec.h:59]: removed

File: src/ingest/SpecRepo.cpp
Comment (chenqin) [reply src/ingest/SpecRepo.cpp:185]: fixed.

File: src/ingest/SpecRepo.cpp
Comment (chenqin) [reply src/ingest/SpecRepo.cpp:139]: fixed.

File: src/ingest/SpecRepo.cpp
Comment (chenqin) [reply src/ingest/SpecRepo.cpp:138]: fixed

File: src/ingest/TimeRow.h
Comment (chenqin) [reply src/ingest/TimeRow.h:146]: fixed

File: src/meta/TableSpec.h
Comment (chenqin) [reply src/meta/TableSpec.h:214]: fixed, will update pr latter today

File: src/meta/TableSpec.h
Comment (chenqin) [reply src/meta/TableSpec.h:216]: agree

File: .gitignore
Comment (shawncao) [inline .gitignore:49]: 👍  which is better compared to VS code?

File: src/ingest/SpecRepo.cpp
Comment (shawncao) [inline src/ingest/SpecRepo.cpp:140]: nit: separate line

File: src/ingest/TimeRow.h
Comment (shawncao) [inline src/ingest/TimeRow.h:30]: thanks!

File: src/meta/TableSpec.h
Comment (shawncao) [inline src/meta/TableSpec.h:210]: nit: can you format this into multiple lines - one entry per line?

File: src/ingest/SpecRepo.cpp
Comment (shawncao) [inline src/ingest/SpecRepo.cpp:188]: why max_days  +2? 
btw, could you rename `max_days` to `maxDays`, it was my mistake?

File: .gitignore
Comment (chenqin) [reply .gitignore:49]: will try learn vs code in 2021.

File: src/ingest/SpecRepo.cpp
Comment (chenqin) [reply src/ingest/SpecRepo.cpp:140]: fixed.

File: src/meta/TableSpec.h
Comment (chenqin) [reply src/meta/TableSpec.h:210]: fixed.

File: src/ingest/SpecRepo.cpp
Comment (chenqin) [reply src/ingest/SpecRepo.cpp:188]: good catch, refactored a bit to use end index explicitly.

File: src/meta/Meta.cmake
Comment (shawncao) [inline src/meta/Meta.cmake:9]: hmm - the indentation doesn't look correct, could you please update tab size to 4 rather than 8 in your IDE and reformat this?

File: src/meta/test/TestTableSpec.cpp
Comment (shawncao) [inline src/meta/test/TestTableSpec.cpp:35]: I think we should enforce macros with format in bracelets "{}", in that sense, I would expect this to be INVALID

File: src/meta/TableSpec.h
Comment (shawncao) [inline src/meta/TableSpec.h:211]: feels to me - these strings should be wrapped by "{}" as "{date}", "{hour}", etc. 
otherwise, too easy to match incorrectly from a path string.

**Review by shawncao — APPROVED** [review]
Looks great so far - just some minor formatting. 
This is an awesome update on macros! Thanks Chen!

File: src/ingest/SpecRepo.cpp
Comment (chenqin) [reply src/ingest/SpecRepo.cpp:156]: I felt I need to test this logic in someway..

File: src/meta/test/TestTableSpec.cpp
Comment (chenqin) [reply src/meta/test/TestTableSpec.cpp:35]: I did quick test using {date} and fill in test.yaml configuration ends with build error, may be we need to handle escape?

File: src/ingest/SpecRepo.cpp
Comment (shawncao) [reply src/ingest/SpecRepo.cpp]: Have this method signature declared in header and then write unit test for it?

File: src/meta/test/TestTableSpec.cpp
Comment (shawncao) [reply src/meta/test/TestTableSpec.cpp:35]: In yaml/yml， you need input encoded url? {date} -> %7Bdate%7D

File: src/meta/TableSpec.h
Comment (chenqin) [reply src/meta/TableSpec.h]: updated in 
c612da3

File: src/ingest/SpecRepo.cpp
Comment (chenqin) [reply src/ingest/SpecRepo.cpp]: added

**Comment by chenqin** [comment]
> why do we introduce `/{ }/` pattern? isn't `{}` enough?

I think if we put {date} in pattern and run test, the test will not run. Might related to cpp-yaml treat { and } as map of configs

**Comment by chenqin** [comment]
> loadConfigured

I see, there are some hard coded logic, could use defined constants in meta::TableSpec  class.
also needs test coverage

File: src/service/test/TestLoadHandler.cpp
Comment (shawncao) [inline src/service/test/TestLoadHandler.cpp:41]: 👍

File: src/service/server/LoadHandler.cpp
Comment (shawncao) [inline src/service/server/LoadHandler.cpp:57]: always use const reference for the parameter to avoid object copy. `const common::unordered_map<>& p`

File: src/meta/TableSpec.h
Comment (shawncao) [inline src/meta/TableSpec.h]: minor for discussion only, I don't have preference: instead of maintain multiple collection due to source format. I'm wondering if we can do single collection without format `/{` or `{`, just purely the MACRO NAME `DATE`, `HOUR`, `MINUTE`, and let the matcher to format the string when doing the search.

Comment (chenqin) [reply src/meta/TableSpec.h]: agree, this seems overkill to avoid those macro , we can put a reserved words list.
like https://ci.apache.org/projects/flink/flink-docs-stable/dev/table/sql/#reserved-keywords

File: src/service/server/LoadHandler.cpp
Comment (chenqin) [reply src/service/server/LoadHandler.cpp:57]: fixed

File: src/meta/TableSpec.h
Comment (chenqin) [reply src/meta/TableSpec.h]: 55e454d use DATE HOUR MINUTE SECOND all cap as macro

Comment (shawncao) [reply src/meta/TableSpec.h]: RESERVED KEYWORD is a good idea when applied in schema/coding, but restrict user arbitrary path (e.g s3) to not have normal words like DATE is too fragile I think. I think in file path, we should still enforce `{DATE}` format, but internally we can treat the MACRO name as reserved `DATE`. what do you think?

**Comment by shawncao** [comment]
@chenqin merge? we can follow up minor things in future PRs.

**Comment by shawncao** [comment]
sorry I forgot to ask - do you have permission to merge (`squash and merge` button)? I approved the PR a few days ago, I thought you can merge any time, but I didn't check if you can do so.

**Comment by chenqin** [comment]
> sorry I forgot to ask - do you have permission to merge (`squash and merge` button)? I approved the PR a few days ago, I thought you can merge any time, but I didn't check if you can do so.

Please add me to collaborator of this project. Meanwhile, please help merge this pr.

**Comment by shawncao** [comment]
> > sorry I forgot to ask - do you have permission to merge (`squash and merge` button)? I approved the PR a few days ago, I thought you can merge any time, but I didn't check if you can do so.
> 
> Please add me to collaborator of this project. Meanwhile, please help merge this pr.

I added you last night, please check your notifications, and let me merge this PR now. 

Congrats on the first MERGE! 🥇

**Pull request merged**
